foundation-model-stack/fms-hf-tuning
# feat: dataclass args for accelerated MoE tuning

### Description of the change
This PR adds one dataclass argument to enable _accelerted moe_ for `sft_trainer.py`, via the new fms-acceleration [accelerated-moe](https://github.com/foundation-model-stack/fms-acceleration/blob/main/plugins/accelerated-moe) plugin and allows for accelerated MoE full-finetuning with the `--fast_moe` flag. `--fast_moe` enables a technique to train Mixture of Expert (MoE) models in parallel instead of sequentially.
With this flag, we expect major speedup in train time and decrease in memory usage on Mixture of Expert models.
| Framework Config                  | EP Degree (parameter) | Model           |   | Train Runtime | Speedup | Memory Usage | Memory Savings |
|-----------------------------------|-----------------------|-----------------|---|---------------|---------|--------------|----------------|
| none                              | N/A                   | granite 3b a800 | 1 | 2371.93       | base    | 71199        | base           |   
| Scatter MoE                       | **1**                     | granite 3b a800 | 1 | 742.739       | **3.19**    | 71187        | **1.0**            |   
| Scatter MoE + Padding Free        | **1**                     | granite 3b a800 | 1 | 631.976       | **3.75**    | 48401        | **0.68**           |   
| Scatter MoE + Padding Free + foak | **1**                     | granite 3b a800 | 1 | 615.453       | **3.85**    | 42651        | **0.6**            |  
| none                              | N/A                   | mixtral 8x7b    | 8 | 4180.95       | base    | 65607        | base           |   
| Scatter MoE                       | **8**                     | mixtral 8b7x    | 8 | 1071.2        | **3.9**     | 52004.8      | **0.79**           |   
| Scatter MoE + Padding Free + foak | **8**                     | mixtral 8x7b    | 8 | 1043.67       | **4.01**    | 51961.2      | **0.79**           |

### Related issue number

### How to verify the PR
**This PR is a work-in-progress and requires more testing, and the official release of `fms-acceleration-moe`**

- To verify, run a tuning job with `fast_moe`.
- Run a tuning job with other plugins added on top of `fast_moe`
- Ensure that incorrect parameters result in failures
- Ensure that non-MoE models cannot be trained with this plugin set

### Was the PR tested
